File: app/src/main/java/com/example/testtanaw/LoginActivity.java
package com.example.testtanaw;

import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.example.testtanaw.adapters.InstitutionSpinnerAdapter;
import com.example.testtanaw.models.Avatar;
import com.example.testtanaw.models.Constants;
import com.example.testtanaw.models.Institution;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.google.firebase.firestore.QuerySnapshot;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;

import java.io.ByteArrayOutputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class LoginActivity extends AppCompatActivity {
    private static final String TAG = "LoginActivity";
    private FirebaseFirestore db;
    private FirebaseStorage storage;
    private FirebaseAuth mAuth;
    private static final int LIMIT = 50;
    private EditText srCodeInput;
    private EditText passwordInput;
    private Spinner institutionSpinner;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        // Initialize Firebase instances
        db = FirebaseFirestore.getInstance();
        mAuth = FirebaseAuth.getInstance();
        storage = FirebaseStorage.getInstance(Constants.BUCKET);

        // Initialize views
        srCodeInput = findViewById(R.id.sr_code_input);
        passwordInput = findViewById(R.id.password_input);

        final List<Institution> institutionList = new ArrayList<>();
        institutionSpinner = findViewById(R.id.institution_spinner);

        // Get institutions from Firestore
        db.collection(Constants.DB_INSTITUTIONS)
                .orderBy("institution", Query.Direction.ASCENDING)
                .limit(LIMIT)
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<QuerySnapshot> task) {
                        if (task.isSuccessful()) {
                            for (QueryDocumentSnapshot document : task.getResult()) {
                                Institution institution = document.toObject(Institution.class);
                                institution.setId(document.getId());
                                institutionList.add(institution);
                            }

                            InstitutionSpinnerAdapter adapter = new InstitutionSpinnerAdapter(
                                    LoginActivity.this,
                                    institutionList
                            );
                            institutionSpinner.setAdapter(adapter);
                        } else {
                            Log.d(TAG, "Error getting documents: ", task.getException());
                        }
                    }
                });
    }

    // Login button click handler
    public void onLoginClicked(View view) {
        String srCode = srCodeInput.getText().toString();
        String password = passwordInput.getText().toString();
        Institution selectedInstitution = (Institution) institutionSpinner.getSelectedItem();

        if (selectedInstitution == null) {
            Toast.makeText(this, "Please select an institution", Toast.LENGTH_SHORT).show();
            return;
        }

        if (srCode.isEmpty() || password.isEmpty()) {
            Toast.makeText(this, "Please fill in all fields", Toast.LENGTH_SHORT).show();
            return;
        }

        String email = srCode + "@" + selectedInstitution.getEmailExtension();

        // Attempt Firebase Authentication
        mAuth.signInWithEmailAndPassword(email, password)
                .addOnCompleteListener(this, task -> {
                    if (task.isSuccessful()) {
                        // Sign in success, get user data and redirect
                        FirebaseUser authUser = mAuth.getCurrentUser();
                        if (authUser != null) {
                            fetchUserDataAndRedirect(authUser.getUid());
                        }
                    } else {
                        // If sign in fails, display a message to the user
                        Toast.makeText(LoginActivity.this, "Authentication failed: " + 
                                task.getException().getMessage(), Toast.LENGTH_SHORT).show();
                    }
                });
    }

    // Open SignUp Activity when the "Sign Up" text is clicked
    public void openSignUp(View view) {
        Intent intent = new Intent(this, SignUpActivity.class);
        startActivity(intent);
    }


    private void fetchUserDataAndRedirect(String userId) {
        db.collection(Constants.DB_USERS).document(userId)
                .get()
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful() && task.getResult() != null) {
                        // Check if the user exists
                        if (task.getResult().exists()) {
                            db.collection(Constants.DB_AVATARS).document(userId)
                                    .get()
                                    .addOnCompleteListener(avatarTask -> {
                                        if (avatarTask.isSuccessful() && avatarTask.getResult() != null) {
                                            // Check if the avatar exists
                                            if (avatarTask.getResult().exists()) {
                                                // Avatar exists, proceed to the main activity
                                                Intent intent = new Intent(LoginActivity.this, MainActivity.class);
                                                startActivity(intent);
                                                finish();
                                            } else {
                                                // Avatar doesn't exist, upload a default one
                                                String path = Avatar.getUserAvatarPath(userId);
                                                Avatar newAvatar = new Avatar(userId, path, "cics", null, null, "boy", "shirt", null, "none");
                                                uploadDefaultAvatar(userId, path);
                                                saveAvatarDataToFirestore(newAvatar);
                                            }
                                        } else {
                                            Toast.makeText(LoginActivity.this, "Failed to fetch user avatar",
                                                    Toast.LENGTH_SHORT).show();
                                        }
                                    });
                        } else {
                            Toast.makeText(LoginActivity.this, "User does not exist",
                                    Toast.LENGTH_SHORT).show();
                        }
                    } else {
                        Toast.makeText(LoginActivity.this, "Failed to fetch user data",
                                Toast.LENGTH_SHORT).show();
                    }
                });
    }


    private byte[] drawableToBitmap() {
        Bitmap bitmap = BitmapFactory.decodeResource(getResources(), R.drawable.avatar);

        // Convert the Bitmap to a byte array
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        bitmap.compress(Bitmap.CompressFormat.PNG, 100, byteArrayOutputStream);
        return byteArrayOutputStream.toByteArray();
    }


    private void uploadDefaultAvatar(String userId, String path) {
        StorageReference storageRef = storage.getReference();
        StorageReference avatarImageRef = storageRef.child(path);

        byte[] data = drawableToBitmap();

        UploadTask uploadTask = avatarImageRef.putBytes(data);
        uploadTask.addOnFailureListener(new OnFailureListener() {
            @Override
            public void onFailure(@NonNull Exception exception) {
                Log.e(TAG, "Avatar image upload failed.", exception);
            }
        }).addOnSuccessListener(new OnSuccessListener<UploadTask.TaskSnapshot>() {
            @Override
            public void onSuccess(UploadTask.TaskSnapshot taskSnapshot) {
                Log.d(TAG, "Avatar image uploaded.");
            }
        });
    }

    private void saveAvatarDataToFirestore(Avatar avatar) {
        // Save data to Firestore under the "avatars" collection
        db.collection(Constants.DB_AVATARS)
                .document(avatar.getUserId())
                .set(avatar)
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful()) {
                        Toast.makeText(LoginActivity.this, "Avatar data saved.", Toast.LENGTH_SHORT).show();

                        // redirect to AvatarActivity
                        Intent intent = new Intent(LoginActivity.this, AvatarActivity.class);
                        startActivity(intent);
                    } else {
                        Toast.makeText(LoginActivity.this, "Failed to save avatar data: " + task.getException().getMessage(),
                                Toast.LENGTH_SHORT).show();
                    }
                });
    }
}

File: app/src/main/java/com/example/testtanaw/LoginActivity.kt
package com.example.testtanaw

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import com.example.testtanaw.models.UserParcelable
import com.example.testtanaw.util.DB
import com.example.testtanaw.util.InstitutionAdapter
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class LoginActivity : AppCompatActivity() {
    private val db = DB()
    private lateinit var srCodeInput: EditText
    private lateinit var passwordInput: EditText
    private lateinit var institutionSpinner: Spinner
    private lateinit var institutions: List<DB.Institution>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)


        CoroutineScope(Dispatchers.Main).launch {
            // Initialize views
            srCodeInput = findViewById(R.id.sr_code_input)
            passwordInput = findViewById(R.id.password_input)
            institutionSpinner = findViewById(R.id.institution_spinner)

            institutions = db.getInstitutions(this@LoginActivity, institutionSpinner)

        }
    }

    // Login button click handler
    fun onLoginClicked(view: View) {
        val srCode = srCodeInput.text.toString()
        val password = passwordInput.text.toString()
        val selectedInstitution = institutions[institutionSpinner.selectedItemPosition].institution
        val emailExtension: String =
            institutions[institutionSpinner.selectedItemPosition].emailExtension

        // Ensure the institution is valid and selected
        if (institutionSpinner.selectedItemPosition == 0) {
            Toast.makeText(this, "Please select a valid institution.", Toast.LENGTH_SHORT).show()
            return
        }

        Log.d("xxxxxx", "$selectedInstitution, $emailExtension")

        CoroutineScope(Dispatchers.Main).launch {
            if (srCode.isNotEmpty() && password.isNotEmpty()) {
                try {
                    val email = srCode + emailExtension
                    Log.d("xxxxxx", "email: $email")

                    val userId = db.login(email, password, this@LoginActivity)
                    if (userId != null) {

                        // check if may avatar -- if wala, redirect sa avatar creation dapat??
                        val userDbData: DB.UserData? = db.getUserDataByUserId(userId);
                        Log.d("xxxxxx", "avatarrr: {${userDbData?.avatars?.get("avatar_url")}}")

                        if (userDbData != null) {
                            if (userDbData.avatars?.get("avatar_url")?.isEmpty() == true ||
                                (userDbData.departments?.get("department")?.isEmpty() == true)
                            ) {
                                // redirect sa create avatar???

                            } else {
                                val userData = UserParcelable(
                                    userDbData.userId,
                                    userDbData.email,
                                    userDbData.srCode,
                                    userDbData.firstName,
                                    userDbData.lastName,
                                    userDbData.institutions["institution"]!!,
                                    userDbData.institutions["institution_logo"]!!,
                                    userDbData.institutions["campus"]!!,
                                    userDbData.departments?.get("department"),
                                    userDbData.avatars?.get("avatar_url")
                                )

                                val intent = Intent(
                                    this@LoginActivity,
                                    MainActivity::class.java
                                )
                                intent.putExtra("userData", userData)
                                startActivity(intent)
                                finish()
                            }
                        }
                    }
                } catch (e: Exception) {
                    Toast.makeText(
                        this@LoginActivity,
                        "Error signing up: ${e.message}",
                        Toast.LENGTH_SHORT
                    ).show()
                }
            } else {
                Toast.makeText(
                    this@LoginActivity,
                    "Please fill in both fields.",
                    Toast.LENGTH_SHORT
                ).show()
            }
        }

    }

    // Open SignUp Activity when the "Sign Up" text is clicked
    fun openSignUp(view: View) {
        val intent = Intent(this, SignUpActivity::class.java)
        startActivity(intent)
    }
}
